### DISCUSSION TITLE
How to pass -fno-sanitize=undefined to `zig cc`?

### DISCUSSION THREAD

**Discussion by jiacai2050** [rust-cross/cargo-zigbuild#224] [Q&A]
Since `zig cc` pass `-fsanitize=undefined -fsanitize-trap=undefined` by default, some project can be built successfully, but
will throw error when running

``` shell
$ ./target/debug/foo
Illegal instruction
```

https://github.com/rust-cross/cargo-zigbuild/blob/871605b68ba6074b24943ec9bdf2b32dea5590c0/src/zig.rs#L510

After some research, it seems there is no way to pass `zig cc` options?

- Ref: https://nathancraddock.com/blog/zig-cc-undefined-behavior/

**Comment by messense** [comment]
Try pass it in `CFLAGS`, `CXXFLAGS` and/or `RUSTFLAGS`.

**Reply by jiacai2050** [reply]
``` shell
$ RUSTFLAGS='-fno-sanitize=undefined' cargo zigbuild

error: failed to run `rustc` to learn about target-specific information

error: Unrecognized option: 'f'


$ CFLAGS='-fno-sanitize=undefined' CXXFLAGS='-fno-sanitize=undefined' cargo zigbuild 
warning: cc: error: unrecognized command line option ‘-fno-sanitize=undefined’

```

No luck.

**Reply by messense** [reply]
* For `RUSTFLAGS` you probably need to pass it in `-C link-arg` since `-f` isn't a rustc option, and target specific `RUSTFLAGS` might be required.
* For `CFLAGS`/`CXXFLAGS` you probably need to pass it in target specific env var (for example `CFLAGS_<target triple>`) since you're cross compilng, you don't want to pass it to build scripts.

**Reply by jiacai2050** [reply]
`RUSTFLAGS` is used for rustc, so I don't think it's the way to go.

It's more likely to be CFLAGS or CXXFLAGS, and I don't use zigbuild for cross compile, just normal build, since my host has old version gcc installed, and my project require newer one.

**Reply by messense** [reply]
> I don't use zigbuild for cross compile, just normal build

It's complicated when build scripts are involved, it's not-so-normal per se.

See

* https://doc.rust-lang.org/nightly/cargo/reference/unstable.html#target-applies-to-host
* https://github.com/PyO3/pyo3/issues/1513

**Reply by jiacai2050** [reply]
I can't imagine the issue is so complex.

```
$ cat ~/bin/zigcpp

zig c++ -fno-sanitize=undefined $@

$ cat ~/bin/zigcc

zig cc -fno-sanitize=undefined $@

$ CC=zigcc CXX=zigcpp zig build
```

I try build like this, it builds and run successfully.  

Since my use case is so simple that I can just use this original solution.

Thanks for your help.

**Reply by jiacai2050** [reply]
FYI, also we need to set linker explicitly like this:

```
[target.x86_64-unknown-linux-gnu]
linker = "zigcc"
```
